sim-lib/src/lib.rs

```diff
@@ -988,64 +1045,79 @@ async fn produce_simulation_results(
     nodes: HashMap<PublicKey, Arc<Mutex<dyn LightningNode>>>,
     mut output_receiver: Receiver<SimulationOutput>,
     results: Sender<(Payment, PaymentResult)>,
-    shutdown: Listener,
-) {
-    log::debug!("Simulation results producer started.");
-
+    listener: Listener,
+) -> Result<(), SimulationError> {
     let mut set = tokio::task::JoinSet::new();
 
     loop {
         tokio::select! {
             biased;
-            _ = shutdown.clone() => break,
+            _ = listener.clone() => {
+                log::debug!("Produce simulation results received shutdown signal. Simulation results producer exiting.");
+                return Ok(())
+            },
             output = output_receiver.recv() => {
-                match output{
+                match output {
                     Some(simulation_output) => {
                         match simulation_output{
                             SimulationOutput::SendPaymentSuccess(payment) => {
-                                let source_node = nodes.get(&payment.source).unwrap().clone();
-                                set.spawn(track_payment_result(
-                                    source_node, results.clone(), payment, shutdown.clone(),
-                                ));
+                                if let Some(source_node) = nodes.get(&payment.source) {
+                                    set.spawn(track_payment_result(
+                                        source_node.clone(), results.clone(), payment, listener.clone()
+                                    ));
```

Comment:
Should we handle errors returned by `track_payment_result` and trigger shutdown here?

Reply:
Yes, we should. I refactored `produce_simulation_results` to include an additional [branch](https://github.com/bitcoin-dev-project/sim-ln/blob/b84ce8d24d5639139fa6bc8ca945557444abd2ff/sim-lib/src/lib.rs#L1087C13-L1102C14) to wait on concurrently. Within this branch, we propagate any `track_payment_result` error to `produce_simulation_results` and trigger shutdown at the latter's spawn site.

Reply:
Nice, realize that we weren't actually waiting on that set at all before this ☠️ 

This method is an interesting one (/different to our other ones) because it has its own set of tasks that it should wait on. As is, if we get the shutdown listener signal we won't wait for all the spawned payment tracking tasks to complete (which is messy). 

Don't need to update in this PR, let's gettit in, but note to self to create an issue/fix this up in future!

Reply:
I've taken note of this and will create an issue for better handling of exits on all tasks spawned in `set`.